apis/python/tests/test_fastercsx.py

```diff
@@ -0,0 +1,370 @@
+    if format == "csr":
+        assert np.array_equal(
+            sp.tocsr()[index or slice(None), :].toarray(), cm.to_numpy(index)
+        )
+    else:
+        assert np.array_equal(
+            sp.tocsc()[:, index or slice(None)].toarray(), cm.to_numpy(index)
+        )
+
+
+def test_bad_arguments(
+    context: soma.SOMATileDBContext, rng: np.random.Generator
+) -> None:
+    """Test various bad argument types/values are caught."""
+    sp = sparse.random(970, 31, density=0.01, dtype=np.float32, random_state=rng)
+
+    with suppress_type_checks():  # w/o this, typeguard raises, which defeats the point of the test
+
+        # context - bad type
+        with pytest.raises(AttributeError):
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row, sp.col, sp.data, sp.shape, "csr", True, None
+            )
+        # context - missing
+        with pytest.raises(TypeError):
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row, sp.col, sp.data, sp.shape, "csr", True
+            )
+
+        # unsupported data type
+        unsup_data = np.full(sp.data.shape, "A", dtype="str")
+        with pytest.raises(ValueError):
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row, sp.col, unsup_data, sp.shape, "csr", True, context
+            )
+
+        # unsupported index type
+        with pytest.raises(ValueError):
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row.astype(np.float32),
+                sp.col.astype(np.float32),
+                sp.data,
+                sp.shape,
+                "csr",
+                True,
+                context,
+            )
+
+        # unsupported index type
+        with pytest.raises(ValueError):
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row.astype(np.int16),
+                sp.col.astype(np.int16),
+                sp.data,
+                sp.shape,
+                "csr",
+                True,
+                context,
+            )
+
+        # mismatched index types
+        with pytest.raises(TypeError):
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row.astype(np.int32),
+                sp.col.astype(np.int64),
+                sp.data,
+                sp.shape,
+                "csr",
+                True,
+                context,
+            )
+
+
+def test_bad_shapes(context: soma.SOMATileDBContext, rng: np.random.Generator) -> None:
+    """Test bad/mismatched shape."""
+    sp = sparse.identity(32, dtype=np.int8).tocoo()
+    for shp in [
+        (sp.shape[0] - 1, sp.shape[1]),
+        (sp.shape[0], sp.shape[1] - 1),
+    ]:
+        with pytest.raises(IndexError):
+            # this will also log an error - which is expected behavior for parallel_for/thread_pool
+            fastercsx.CompressedMatrix.from_ijd(
+                sp.row, sp.col, sp.data, shp, "csr", True, context
+            )
+
+
+"""
+other tests to consider
```

Comment:
Implement them, or remove this comment, and/or file a tracking issue for follow-up work?